### PR TITLE
feat(api): Configurable response time badge thresholds

### PR DESCRIPTION
**Pull request by Jesibu**
## Summary
view original pull request for more information #267 

## Checklist

- [X] Tested and/or added tests to validate that the changes work as intended, if applicable.
- [X] Added the documentation in `README.md`, if applicable.

### CODE REVIEW

File: controller/handler/badge.go [hunk 1]
```diff
@@ -199,7 +206,7 @@ func getBadgeColorFromUptime(uptime float64) string {
 	return badgeColorHexVeryBad
 }
 
-func generateResponseTimeBadgeSVG(duration string, averageResponseTime int) []byte {
+func generateResponseTimeBadgeSVG(duration string, averageResponseTime int, key string, config *config.Config) []byte {
```
Comment (TwiN) [inline controller/handler/badge.go:209]: Can you rename `config *config.Config` to `cfg *config.Config` to avoid package name collision?

File: controller/handler/badge.go [hunk 2]
```diff
@@ -247,17 +254,14 @@ func generateResponseTimeBadgeSVG(duration string, averageResponseTime int) []by
...
-		return badgeColorHexAwesome
-	} else if responseTime <= 200 {
-		return badgeColorHexGreat
-	} else if responseTime <= 300 {
-		return badgeColorHexGood
-	} else if responseTime <= 500 {
-		return badgeColorHexPassable
-	} else if responseTime <= 750 {
-		return badgeColorHexBad
+func getBadgeColorFromResponseTime(responseTime int, key string, config *config.Config) string {
```
Comment (TwiN) [inline controller/handler/badge.go:257]: Can you rename `config *config.Config` to `cfg *config.Config` to avoid package name collision?

File: controller/handler/badge_test.go [hunk 3]
```diff
@@ -11,9 +11,10 @@ import (
 	"github.com/TwiN/gatus/v4/core"
 	"github.com/TwiN/gatus/v4/storage/store"
 	"github.com/TwiN/gatus/v4/watchdog"
+	"github.com/TwiN/gatus/v4/core/ui"
 )
 
-func TestBadge(t *testing.T) {
+func TestUptimeBadge(t *testing.T) {
```
Comment (TwiN) [inline controller/handler/badge_test.go:17]: Can you revert this change? It doesn't only test uptime badges.

File: core/ui/ui.go [hunk 4]
```diff
@@ -1,13 +1,45 @@
...
 // Config is the UI configuration for core.Endpoint
 type Config struct {
 	// HideHostname whether to hide the hostname in the Result
 	HideHostname bool `yaml:"hide-hostname"`
 	// HideURL whether to ensure the URL is not displayed in the results. Useful if the URL contains a token.
 	HideURL bool `yaml:"hide-url"`
 	// DontResolveFailedConditions whether to resolve failed conditions in the Result for display in the UI
-	DontResolveFailedConditions bool `yaml:"dont-resolve-failed-conditions"`
+	DontResolveFailedConditions bool   `yaml:"dont-resolve-failed-conditions"`
+	Badge                       *Badge `yaml:"badge"`
```
Comment (TwiN) [inline core/ui/ui.go:13]:
```suggestion
	// Badge is the configuration for the badges generated
	Badge *Badge `yaml:"badge"`
```

File: core/ui/ui.go [hunk 5]
```diff
@@ -1,13 +1,45 @@
...
+		for i := 4; i > 0; i-- {
+			if config.Badge.ResponseTime.Thresholds[i] < config.Badge.ResponseTime.Thresholds[i-1] {
+				return ErrInvalidBadgeResponseTimeConfig
+			}
+		}
+		//config.Badge.ResponseTime = GetDefaultConfig().Badge.ResponseTime
+	} else {
+		config.Badge = GetDefaultConfig().Badge
+	}
+
```
Comment (TwiN) [inline core/ui/ui.go:41]:
```suggestion
	} else {
		config.Badge = GetDefaultConfig().Badge
	}
```

**Review by TwiN — CHANGES_REQUESTED** [review]
A few small changes and we should be good to go!

File: controller/handler/badge.go (hunk 2 above)
Comment (Jesibu) [reply controller/handler/badge.go:257]: done

File: controller/handler/badge.go [hunk 6]
```diff
@@ -247,17 +254,14 @@ func generateResponseTimeBadgeSVG(duration string, averageResponseTime int) []by
...
-	} else if responseTime <= 200 {
-		return badgeColorHexGreat
-	} else if responseTime <= 300 {
-		return badgeColorHexGood
-	} else if responseTime <= 500 {
-		return badgeColorHexPassable
-	} else if responseTime <= 750 {
-		return badgeColorHexBad
+func getBadgeColorFromResponseTime(responseTime int, key string, cfg *config.Config) string {
+
```
Comment (TwiN) [inline controller/handler/badge.go:258]:
```suggestion
```

File: README.md [hunk 7]
```diff
@@ -1469,6 +1470,23 @@ Where:
 - `{duration}` is `7d`, `24h` or `1h`
 - `{key}` has the pattern `<GROUP_NAME>_<ENDPOINT_NAME>` in which both variables have ` `, `/`, `_`, `,` and `.` replaced by `-`.
 
+##### How to change the color thresholds of the response time badge  
+To change the response time badges threshold, a corresponding configuration can be added to an endpoint.   
+The values in the array correspond to the levels [Awesome, Great, Good, Passable, Bad]  
+All five values must be given in milliseconds (ms).  
+
+```
+- name: nas
```
Comment (TwiN) [inline README.md:1479]:
```suggestion
endpoints:
- name: nas
```

File: controller/handler/badge_test.go [hunk 8]
```diff
@@ -180,7 +211,68 @@ func TestGetBadgeColorFromUptime(t *testing.T) {
...
+				Group:    "group",
+				UIConfig: ui.GetDefaultConfig(),
+			},
+		},
+	}
+
+	testEndpoint.UIConfig = ui.GetDefaultConfig()
+	store.Get().Insert(&testEndpoint, &testSuccessfulResult)
+	cfg.Endpoints[0] = &testEndpoint
+
```
Comment (TwiN) [inline controller/handler/badge_test.go:275]: I'm a little confused here about what you're trying to do. 

Why don't you just set testEndpoint's `UIConfig` field at L234 instead, and then set cfg's `Endpoints` to `testEndpoints`?

Essentially, you'd have something like this:

```go
	cfg := &config.Config{
		Metrics:   true,
		Endpoints: []*core.Endpoint{&testEndpoint},
	}
```
and you wouldn't need to have either L272 and L274

**Review by TwiN — CHANGES_REQUESTED** [review]

File: controller/handler/badge_test.go (hunk 8 above)
Comment (Jesibu) [reply controller/handler/badge_test.go:275]: I actually didn't know what I did there I only transfered this from the old branch. I implemented your suggestions

**Review by TwiN — APPROVED** [review]
Sounds good, thank you for the contribution!

**Comment by Jesibu** [comment]
Was fun. Thanks for your advise
